### PR TITLE
Add support for DetectoRS.

### PR DESCRIPTION
**Pull request by joe-siyuan-qiao**
Dear developers,

This PR adds support for DetectoRS. I left the table in README.md empty. The corresponding pre-trained models and the training logs can be found here:

https://github.com/joe-siyuan-qiao/DetectoRS/tree/mmdetv2

I patched the diff from the above implementation and re-tested every pre-trained models. The results are consistent. I didn't re-train them using the codes submitted here.

Thanks a lot for developing mmdetection!

Cheers!

### CODE REVIEW

**Comment by xvjiarui** [comment]
Hi @joe-siyuan-qiao 
Thanks very much for your contribution!
I will review and benchmark this PR. 

Could you please fix the CI first?

**Comment by joe-siyuan-qiao** [comment]
The issues have been fixed. Thanks!

File: configs/detectors/cascade_rcnn_r50_rfp_1x_coco.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+            type='ResNet',
+            depth=50,
+            num_stages=4,
+            out_indices=(0, 1, 2, 3),
+            frozen_stages=1,
+            norm_cfg=dict(type='BN', requires_grad=True),
+            norm_eval=True,
+            conv_cfg=dict(type='ConvAWS'),
+            style='pytorch'),
+        rfp_pretrained='torchvision://resnet50',
```
Comment (ZwwWayne) [inline configs/detectors/cascade_rcnn_r50_rfp_1x_coco.py:23]: ',' is unnecessary.

File: configs/detectors/detectors_cascade_rcnn_r50_1x_coco.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            num_stages=4,
+            out_indices=(0, 1, 2, 3),
+            frozen_stages=1,
+            norm_cfg=dict(type='BN', requires_grad=True),
+            norm_eval=True,
+            conv_cfg=dict(type='ConvAWS'),
+            sac=dict(type='SAC', use_deform=True),
+            stage_with_sac=(False, True, True, True),
+            style='pytorch'),
+        rfp_pretrained='torchvision://resnet50',
```
Comment (ZwwWayne) [inline configs/detectors/detectors_cascade_rcnn_r50_1x_coco.py:29]: ',' is unnecessary.

File: mmdet/models/necks/rfp.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import xavier_init
+
+from ..builder import NECKS, build_backbone
+from .fpn import FPN
+
+
+class ASPP(torch.nn.Module):
+
```
Comment (ZwwWayne) [inline mmdet/models/necks/rfp.py:13]: Docstring.

File: mmdet/models/necks/rfp.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    def __init__(self, in_channels, out_channels):
+        super().__init__()
+        kernel_sizes = [1, 3, 3, 1]
```
Comment (ZwwWayne) [inline mmdet/models/necks/rfp.py:16]: Make them configurable parameters.

File: mmdet/ops/conv_ws.py [hunk 5]
```diff
@@ -46,3 +47,71 @@ def __init__(self,
     def forward(self, x):
         return conv_ws_2d(x, self.weight, self.bias, self.stride, self.padding,
                           self.dilation, self.groups, self.eps)
+
+
+@CONV_LAYERS.register_module('ConvAWS')
+class ConvAWS2d(nn.Conv2d):
+
```
Comment (ZwwWayne) [inline mmdet/ops/conv_ws.py:54]: Docstring.

File: mmdet/ops/conv_ws.py [hunk 6]
```diff
@@ -46,3 +47,71 @@ def __init__(self,
...
+            padding=padding,
+            dilation=dilation,
+            groups=groups,
+            bias=bias)
+        self.register_buffer('weight_gamma',
+                             torch.ones(self.out_channels, 1, 1, 1))
+        self.register_buffer('weight_beta',
+                             torch.zeros(self.out_channels, 1, 1, 1))
+
+    def _get_weight(self, weight):
```
Comment (ZwwWayne) [inline mmdet/ops/conv_ws.py:78]: Please simplify the code and refer to the implementation of [ConvWS](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/ops/conv_ws.py#L14).

File: mmdet/ops/saconv.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        self.pre_context = torch.nn.Conv2d(
+            self.in_channels, self.in_channels, kernel_size=1, bias=True)
+        self.pre_context.weight.data.fill_(0)
+        self.pre_context.bias.data.fill_(0)
+        self.post_context = torch.nn.Conv2d(
+            self.out_channels, self.out_channels, kernel_size=1, bias=True)
+        self.post_context.weight.data.fill_(0)
+        self.post_context.bias.data.fill_(0)
+        if self.use_deform:
+            self.offset_s = torch.nn.Conv2d(
```
Comment (ZwwWayne) [inline mmdet/ops/saconv.py:69]: torch.nn.xxx -> nn.xxx

File: mmdet/ops/saconv.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                stride=stride,
+                bias=True)
+            self.offset_s.weight.data.fill_(0)
+            self.offset_s.bias.data.fill_(0)
+            self.offset_l.weight.data.fill_(0)
+            self.offset_l.bias.data.fill_(0)
+
+    def forward(self, x):
+        # pre-context
+        avg_x = torch.nn.functional.adaptive_avg_pool2d(x, output_size=1)
```
Comment (ZwwWayne) [inline mmdet/ops/saconv.py:90]: import torch.nn as nn
import nn.functional as F

File: mmdet/models/backbones/resnet.py [hunk 9]
```diff
@@ -293,6 +323,51 @@ def _inner_forward(x):
...
+            rfp_feat = self.rfp_conv(rfp_feat)
+            out = out + rfp_feat
+
+        out = self.relu(out)
+
+        return out
+
 
 @BACKBONES.register_module()
 class ResNet(nn.Module):
```
Comment (xvjiarui) [inline mmdet/models/backbones/resnet.py:373]: We may inherit `ResNet` to make the structure simpler.

File: mmdet/models/backbones/resnet.py [hunk 10]
```diff
@@ -210,6 +229,17 @@ def __init__(self,
...
 
+        self.rfp_inp = rfp_inp
+        if self.rfp_inp:
+            self.rfp_conv = nn.Conv2d(
+                self.rfp_inp,
+                planes * self.expansion,
+                kernel_size=1,
+                stride=1,
+                bias=True)
+            self.rfp_conv.weight.data.fill_(0)
```
Comment (xvjiarui) [inline mmdet/models/backbones/resnet.py:240]: We may use init func from `mmcv.cnn`.

File: mmdet/models/backbones/resnet.py [hunk 11]
```diff
@@ -210,6 +229,17 @@ def __init__(self,
             self.after_conv3_plugin_names = self.make_block_plugins(
                 planes * self.expansion, self.after_conv3_plugins)
 
+        self.rfp_inp = rfp_inp
+        if self.rfp_inp:
```
Comment (xvjiarui) [inline mmdet/models/backbones/resnet.py:233]: `rfp_inplanes` may be better?

File: mmdet/models/backbones/resnet.py [hunk 12]
```diff
@@ -210,6 +229,17 @@ def __init__(self,
...
+            self.rfp_conv = nn.Conv2d(
```
Comment (xvjiarui) [inline mmdet/models/backbones/resnet.py:234]: We may replace `nn.Conv2d` with `build_conv_layer`.

File: mmdet/models/backbones/resnet.py [hunk 13]
```diff
@@ -615,6 +702,26 @@ def forward(self, x):
             x = res_layer(x)
             if i in self.out_indices:
                 outs.append(x)
+        if self.output_img:
+            outs.insert(0, img)
```
Comment (xvjiarui) [inline mmdet/models/backbones/resnet.py:706]: We may resolve this hack by inheritance.

File: mmdet/ops/conv_ws.py [hunk 14]
```diff
@@ -46,3 +47,71 @@ def __init__(self,
...
+        super()._load_from_state_dict(state_dict, prefix, local_metadata,
+                                      strict, missing_keys, unexpected_keys,
+                                      error_msgs)
+        if self.weight_gamma.data.mean() > 0:
+            return
+        weight = self.weight.data
+        weight_mean = weight.data.mean(
+            dim=1, keepdim=True).mean(
+                dim=2, keepdim=True).mean(
+                    dim=3, keepdim=True)
```
Comment (xvjiarui) [inline mmdet/ops/conv_ws.py:107]: Similarly as above.

File: mmdet/ops/saconv.py (hunk 7 above)
Comment (xvjiarui) [inline mmdet/ops/saconv.py:69]: nn.Conv2d

**Comment by xvjiarui** [comment]
Hi @joe-siyuan-qiao 
Thanks for your prompt response. 
There are still some unresolved comments, e.g. replace `torch.nn` with `nn`, make a new class `ResNeSt`.

File: mmdet/models/backbones/resnet.py [hunk 15]
```diff
@@ -25,9 +25,13 @@ def __init__(self,
                  conv_cfg=None,
                  norm_cfg=dict(type='BN'),
                  dcn=None,
+                 rfp_inplanes=None,
+                 sac=None,
```
Comment (xvjiarui) [inline mmdet/models/backbones/resnet.py:29]: If we make a new class `ResNeSt`, we don't need to support `BasicBlock`

File: mmdet/models/backbones/resnet.py [hunk 16]
```diff
@@ -210,6 +229,17 @@ def __init__(self,
...
+        self.rfp_inplanes = rfp_inplanes
+        if self.rfp_inplanes:
+            self.rfp_conv = build_conv_layer(
+                None,
+                self.rfp_inplanes,
+                planes * self.expansion,
+                1,
+                stride=1,
+                bias=True)
+            constant_init(self.rfp_conv, 0)
```
Comment (xvjiarui) [inline mmdet/models/backbones/resnet.py:241]: We may move it into `init_weights()`.

File: mmdet/models/backbones/resnet.py [hunk 17]
```diff
@@ -372,6 +447,10 @@ def __init__(self,
                  norm_eval=True,
                  dcn=None,
                  stage_with_dcn=(False, False, False, False),
+                 sac=None,
+                 stage_with_sac=(False, False, False, False),
+                 rfp_inplanes=None,
+                 output_img=False,
```
Comment (xvjiarui) [inline mmdet/models/backbones/resnet.py:453]: We may add docstring about these arguments.

File: mmdet/models/necks/rfp.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+        dilations (list): Dilations of the four branches
+        paddings (list): Paddings of the four branches
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 kernel_sizes=[1, 3, 3, 1],
+                 dilations=[1, 3, 6, 1],
+                 paddings=[0, 3, 6, 0]):
```
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:31]: We may change the default value to tuple, which is unmutable.

File: mmdet/models/necks/rfp.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+
+    This is an implementation of the ASPP module used in DetectoRS
+    (https://arxiv.org/pdf/2006.02334.pdf)
+
+    Args:
+        in_channels (int): Number of channels in the input
+        out_channels (int): Number of channels produced by this module
+        kernel_sizes (list): Kernel sizes of the four branches
+        dilations (list): Dilations of the four branches
+        paddings (list): Paddings of the four branches
```
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:23]: We may change it to tuple and add the following:
`Default: (1, 3, 3, 1)`.

File: mmdet/models/necks/rfp.py [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+        for m in self.modules():
+            if isinstance(m, torch.nn.Conv2d):
+                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
+                m.weight.data.normal_(0, math.sqrt(2. / n))
+                m.bias.data.fill_(0)
+
+    def forward(self, x):
+        avg_x = self.gap(x)
+        out = []
+        for aspp_idx in range(self.aspp_num):
```
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:55]: `self.aspp_num` is not necessary. 
We may directly iterate `self.aspp`.

File: mmdet/models/necks/rfp.py [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+            conv = torch.nn.Conv2d(
+                in_channels,
+                out_channels,
+                kernel_size=kernel_sizes[aspp_idx],
+                stride=1,
+                dilation=dilations[aspp_idx],
+                padding=paddings[aspp_idx],
+                bias=True)
+            self.aspp.append(conv)
+        self.gap = torch.nn.AdaptiveAvgPool2d(1)
```
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:44]: `torch.nn` -> `nn`.

File: mmdet/models/necks/rfp.py [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+        self.aspp_num = len(kernel_sizes)
+        for m in self.modules():
+            if isinstance(m, torch.nn.Conv2d):
+                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
+                m.weight.data.normal_(0, math.sqrt(2. / n))
+                m.bias.data.fill_(0)
```
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:50]: We may wrap these lines into `init_weights` and call them here. 
Also, we may use init functions in `mmcv.cnn` like `kaiming_init`.

File: mmdet/models/necks/rfp.py [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+        super().__init__()
+        self.aspp = torch.nn.ModuleList()
+        for aspp_idx in range(len(kernel_sizes)):
+            conv = torch.nn.Conv2d(
+                in_channels,
+                out_channels,
+                kernel_size=kernel_sizes[aspp_idx],
+                stride=1,
+                dilation=dilations[aspp_idx],
+                padding=paddings[aspp_idx],
```
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:41]: We may only leave `dilations` configurable. 
```
padding = dilation if dilation >= 1 else dilation
kernel_size = 3 if dilation > 1 else 1
```

File: mmdet/models/necks/rfp.py [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+
+    Args:
+        rfp_steps (int): Number of unrolled steps of RFP
+        rfp_backbone (dict): Configuration of the backbone
+        rfp_pretrained (str): Specifies the pre-trained models
+            for rfp_backbone
+        kwargs: Args for FPN
+    """
+
+    def __init__(self, rfp_steps, rfp_backbone, rfp_pretrained, **kwargs):
```
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:78]: We may add `pool_scales` in the `__init__` arguments.

File: mmdet/models/necks/rfp.py [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+        self.rfp_aspp = ASPP(neck_out_channels, neck_out_channels // 4)
+        self.rfp_weight = nn.Conv2d(
+            neck_out_channels,
+            1,
+            kernel_size=1,
+            stride=1,
+            padding=0,
+            bias=True)
+        self.rfp_weight.weight.data.fill_(0)
+        self.rfp_weight.bias.data.fill_(0)
```
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:96]: We may use init functions in `mmcv.cnn`.

File: mmdet/ops/saconv.py [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+                self.in_channels,
+                18,
+                kernel_size=3,
+                padding=1,
+                stride=stride,
+                bias=True)
+            self.offset_s.weight.data.fill_(0)
+            self.offset_s.bias.data.fill_(0)
+            self.offset_l.weight.data.fill_(0)
+            self.offset_l.bias.data.fill_(0)
```
Comment (xvjiarui) [inline mmdet/ops/saconv.py:88]: We may use init functions in `mmcv.cnn`.

File: mmdet/ops/saconv.py [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+        self.switch.bias.data.fill_(1)
+        self.weight_diff = nn.Parameter(torch.Tensor(self.weight.size()))
+        self.weight_diff.data.zero_()
+        self.pre_context = nn.Conv2d(
+            self.in_channels, self.in_channels, kernel_size=1, bias=True)
+        self.pre_context.weight.data.fill_(0)
+        self.pre_context.bias.data.fill_(0)
+        self.post_context = nn.Conv2d(
+            self.out_channels, self.out_channels, kernel_size=1, bias=True)
+        self.post_context.weight.data.fill_(0)
```
Comment (xvjiarui) [inline mmdet/ops/saconv.py:68]: We may use init functions in `mmcv.cnn`.

File: mmdet/models/necks/rfp.py (hunk 22 above)
Comment (xvjiarui) [inline mmdet/models/necks/rfp.py:50]: We may use init func in `mmcv.cnn`.

**Comment by xvjiarui** [comment]
Hi @joe-siyuan-qiao 
After discussing with @hellock, we propose the following modifications to make DetectoRS more versatile. 
1. We may dd a new backbone `RFPResNet` or sth you think more precise and straight forward, which includes `rfp_forward`, `output_img`, `SAC`. For `output_img`, we could make it an option for `RFPResNet`, if `output_img == True`, we directly return `img` and all levels feature map together.  
2. We could make `pretrained` an attribute of the backbone. In this way, we don't need to pass `pretrained` into `RFP`. 

Your contribution is truly appreciated. Cheers!

**Comment by joe-siyuan-qiao** [comment]
Hi @xvjiarui 

Thanks for your comments! I have revised the codes accordingly. Please take a look when you have time. Thanks!

Cheers!

File: mmdet/models/backbones/detectors_resnet.py [hunk 28]
```diff
@@ -0,0 +1,190 @@
+import torch.utils.checkpoint as cp
+from mmcv.cnn import build_conv_layer, constant_init
+
+from ..builder import BACKBONES
+from .resnet import Bottleneck as _Bottleneck
+from .resnet import ResNet
+
+
+class Bottleneck(_Bottleneck):
+    expansion = 4
```
Comment (ZwwWayne) [inline mmdet/models/backbones/detectors_resnet.py:10]: Please add standard docstrings for newly added modules and functions. MMDet V2.0 has more strict requirements on the docstrings.

File: mmdet/ops/conv_ws.py [hunk 29]
```diff
@@ -46,3 +47,85 @@ def __init__(self,
...
+        return weight
+
+    def forward(self, x):
+        weight = self._get_weight(self.weight)
+        return F.conv2d(x, weight, self.bias, self.stride, self.padding,
+                        self.dilation, self.groups)
+
+    def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                              missing_keys, unexpected_keys, error_msgs):
+        self.weight_gamma.data.fill_(-1)
```
Comment (ZwwWayne) [inline mmdet/ops/conv_ws.py:114]: We need docstrings and more comments about this function, e.g., what is the difference of the overrode `_load_from_state_dict` from the original one and why it is necessary.

File: mmdet/ops/saconv.py [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+            kernel_size,
+            stride=stride,
+            padding=padding,
+            dilation=dilation,
+            groups=groups,
+            bias=bias)
+        self.use_deform = use_deform
+        self.switch = nn.Conv2d(
+            self.in_channels, 1, kernel_size=1, stride=stride, bias=True)
+        self.switch.weight.data.fill_(0)
```
Comment (xvjiarui) [inline mmdet/ops/saconv.py:58]: We may use init function in `mmcv.cnn`.

**Comment by xvjiarui** [comment]
Hi @joe-siyuan-qiao 
I made some minor refactoring. You may check if the modification is reasonable and acceptable for you.
Btw, I also benchmarked the models, and [README.md](https://github.com/open-mmlab/mmdetection/blob/9bb993ade28fce9d7099ed0bbe4d737b3a19b6a0/configs/detectors/README.md) is updated.

**Comment by joe-siyuan-qiao** [comment]
Hi @xvjiarui 

Thanks a lot for your efforts! The codes look great to me. 👍 

The only thing that worries me a little bit is that DetectoRS with Cascade + ResNet-50 isn't very close to what I reported. The other results look fine. I'll try to see myself if there are any errors or implementation differences that caused it. But that will be put offline and let's save it for another PR if necessary.

Thanks again!

Cheers!

File: mmdet/models/backbones/detectors_resnext.py [hunk 31]
```diff
@@ -0,0 +1,121 @@
...
+        self.groups = groups
+        self.base_width = base_width
+        super(DetectoRS_ResNeXt, self).__init__(**kwargs)
+
+    def make_res_layer(self, **kwargs):
+        return super().make_res_layer(
+            groups=self.groups,
+            base_width=self.base_width,
+            base_channels=self.base_channels,
+            **kwargs)
```
Comment (hellock) [inline mmdet/models/backbones/detectors_resnext.py:121]: Are `__init__` and `make_res_layer` necessary?

File: mmdet/models/necks/rfp.py [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+        out_channels (int): Number of channels produced by this module
+        dilations (tuple[int]): Dilations of the four branches.
+            Default: (1, 3, 6, 1)
+    """
+
+    def __init__(self, in_channels, out_channels, dilations=(1, 3, 6, 1)):
+        super().__init__()
+        assert dilations[-1] == 1
+        self.aspp = nn.ModuleList()
+        for aspp_idx in range(len(dilations)):
```
Comment (hellock) [inline mmdet/models/necks/rfp.py:27]: `for dilation in dilations`

File: mmdet/models/necks/rfp.py [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+        self.rfp_aspp = ASPP(self.out_channels, aspp_out_channels,
+                             aspp_dilations)
+        self.rfp_weight = nn.Conv2d(
+            self.out_channels,
+            1,
+            kernel_size=1,
+            stride=1,
+            padding=0,
+            bias=True)
+        constant_init(self.rfp_weight, 0)
```
Comment (hellock) [inline mmdet/models/necks/rfp.py:97]: This line can be moved to `init_weights()`.

File: mmdet/ops/saconv.py [hunk 34]
```diff
@@ -0,0 +1,122 @@
...
+                bias=True)
+            self.offset_l = nn.Conv2d(
+                self.in_channels,
+                18,
+                kernel_size=3,
+                padding=1,
+                stride=stride,
+                bias=True)
+            constant_init(self.offset_s, 0)
+            constant_init(self.offset_l, 0)
```
Comment (hellock) [inline mmdet/ops/saconv.py:83]: I suggest defining a `init_weights` method and call it at the end of `__init__`.

File: tools/test.py [hunk 35]
```diff
@@ -87,6 +87,10 @@ def main():
     if cfg.get('cudnn_benchmark', False):
         torch.backends.cudnn.benchmark = True
     cfg.model.pretrained = None
+    if cfg.model.get('neck', False):
```
Comment (hellock) [inline tools/test.py:90]: `cfg.model.get('neck')` will return None if neck does not exist, thus we can omit the default value.

**Comment by xvjiarui** [comment]
Hi @joe-siyuan-qiao 
The PR is merged! Your contribution is truly appreciated.

**Comment by yutao007** [comment]
谢谢作者。不过现在只有版本1能够成功运行此算法。版本2,i还不行。另外请教作者一件事，我将backbone变成res2net。版本1不支持。按照论文的话，应该是可以换backbone的吧！

**Pull request merged**